lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -263,3 +264,144 @@ void mlir::vector_ext::populatePropagateVectorDistributionPatterns(
+  SmallVector<Value> bbArgReplacements;
+  for (const auto &it : llvm::enumerate(warpOp.args())) {
+    Value val = it.value();
+    Value bbArg = warpOpBody->getArgument(it.index());
+
+    // Store arg vector into buffer.
+    rewriter.setInsertionPoint(ifOp);
+    auto vectorType = val.getType().cast<VectorType>();
+    int64_t storeSize = vectorType.getShape()[0];
+    Value storeOffset = rewriter.create<arith::AddIOp>(loc, loadOffset,
+        rewriter.create<arith::MulIOp>(loc, warpOp.laneid(),
+            rewriter.create<arith::ConstantIndexOp>(loc, storeSize)));
+    rewriter.create<vector::StoreOp>(loc, val, buffer, storeOffset);
+
+    // Load bbArg vector from buffer.
+    rewriter.setInsertionPointToStart(ifOp.thenBlock());
+    auto bbArgType = bbArg.getType().cast<VectorType>();
+    Value loadOp = rewriter.create<vector::LoadOp>(loc, bbArgType, buffer,
+                                                   loadOffset);
+    bbArgReplacements.push_back(loadOp);
+
+    // Compute next offset.
+    rewriter.setInsertionPoint(ifOp);
+    loadOffset = rewriter.create<arith::AddIOp>(loc, loadOffset,
+        rewriter.create<arith::ConstantIndexOp>(loc, bbArgType.getShape()[0]));
+  }
+
+  // Move body of warpOp to ifOp.
+  rewriter.mergeBlocks(warpOpBody, ifOp.thenBlock(), bbArgReplacements);
+
+  // Rewrite terminator and compute replacements of WarpOp results.
+  SmallVector<Value> replacements;
+  auto yieldOp = cast<vector_ext::YieldOp>(ifOp.thenBlock()->getTerminator());
+  Location yieldLoc = yieldOp.getLoc();
+  Value storeOffset = c0;
+
+  for (const auto &it : llvm::enumerate(yieldOp.operands())) {
+    Value val = it.value();
+
+    // Store yielded vector into buffer.
+    rewriter.setInsertionPoint(yieldOp);
+    auto vectorType = val.getType().cast<VectorType>();
+    rewriter.create<vector::StoreOp>(yieldLoc, val, buffer, storeOffset);
+
+    // Load vector from buffer (after warpOp).
+    rewriter.setInsertionPointAfter(ifOp);
+    auto loadedVectorType =
+        warpOp->getResultTypes()[it.index()].cast<VectorType>();
+    int64_t loadSize = loadedVectorType.getShape()[0];
+
+    // loadOffset = storeOffset + laneid * loadSize
+    Value loadOffset = rewriter.create<arith::AddIOp>(loc, storeOffset,
+        rewriter.create<arith::MulIOp>(loc, warpOp.laneid(),
+            rewriter.create<arith::ConstantIndexOp>(loc, loadSize)));
+    Value loadOp = rewriter.create<vector::LoadOp>(loc, loadedVectorType,
+                                                   buffer, loadOffset);
+    replacements.push_back(loadOp);
+
+    // Compute next offset.
+    rewriter.setInsertionPoint(ifOp);
+    Value vectorSize = rewriter.create<arith::ConstantIndexOp>(
+        yieldLoc, vectorType.getShape()[0]);
+    storeOffset = rewriter.create<arith::AddIOp>(yieldLoc, storeOffset,
+                                                 vectorSize);
+  }
+
+  // Delete terminator and add empty scf.yield.
+  rewriter.eraseOp(yieldOp);
```

Comment:
nit: is this needed since we are replacing the whole warpOp anyway?

Reply:
The yieldOp was moved over to the `scf.if` by `mergeBlocks` and must be deleted.